sui-execution/latest/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -685,6 +710,18 @@ mod checked {
                 "Events should be taken after every Move call"
             );
 
+            for id in by_value_shared_objects {
+                if !writes.contains_key(&id)
+                    && !deleted_object_ids.contains_key(&id)
+                    && !external_transfers.contains(&id)
```

Comment:
What are the cases that we are allowing with `!writes.contains_key(&id)` and `!external_transfers.contains(&id)`? i.e. if we pass a shared object by value, are we allowed to do anything else other than deleting it?

Reply:
> but is eventually deleted or reshared by the end of the txn

If it is eventually deleted, it should show up in `deleted_object_ids`?
As for re-share, is there another way to share an object other than calling `share_object` or `public_share_object`? If not, I assumed that we would abort with `ESharedNonNewObject` if the object being shared was not created in this transaction.

Reply:
That was changed in this PR to allow it to be reshared (since that wasn't something we had to consider before). It seems helpful for code that conditionally deletes a shared object

Reply:
This logic is sensitive and was really difficult to get right, I don't want to change it. If @lxfind wants further explanation we can chat

Reply:
Could you add comments to the code to explain what each check covers then?